lib/CliftTransforms/TerminalBranchComplementHoisting.cpp

```diff
@@ -0,0 +1,166 @@
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include "revng/Clift/Clift.h"
+#include "revng/CliftTransforms/BooleanNegations.h"
+#include "revng/CliftTransforms/Passes.h"
+#include "revng/CliftTransforms/RewriteHelpers.h"
+
+namespace mlir {
+namespace clift {
+#define GEN_PASS_DEF_CLIFTTERMINALBRANCHCOMPLEMENTHOISTING
+#include "revng/CliftTransforms/Passes.h.inc"
+} // namespace clift
+} // namespace mlir
+
+namespace clift = mlir::clift;
+using namespace clift;
+
+namespace {
+
+enum class HoistingTarget : bool {
+  Then,
+  Else,
+};
+
+static bool hasSingleNoFallthrough(mlir::Region &R) {
+  return static_cast<bool>(getLastNoFallthroughStatement(R));
+}
+
+// Computes an approximation of the size of a statement region in C.
+static double approximateRegionWeight(mlir::Region &R) {
+  revng_assert(R.hasOneBlock());
+
+  double Weight = 0.0;
+  R.walk([&Weight](mlir::Operation *Op) {
+    // Label declarations can be ignored, as they have no C representation.
+    if (mlir::isa<MakeLabelOp>(Op))
+      return;
+
+    // Neither expression statement nor yield operations have a direct C
+    // representation.
+    if (mlir::isa<ExpressionStatementOp, YieldOp>(Op))
+      return;
+
+    Weight += 1.0;
+  });
+  return Weight;
+}
+
+// Attempts to select one branch of an if-statement to be inlined into the
+// nesting scope. If neither branch should be hoisted, the result is nullopt.
+static std::optional<HoistingTarget> selectHoistingTarget(IfOp If) {
+  if (If.getElse().empty())
+    return std::nullopt;
+
+  if (If.getThen().empty())
+    return HoistingTarget::Then;
+
+  if (If.getElse().front().empty())
+    return HoistingTarget::Else;
+
+  bool ThenFallthrough = not isIndirectlyNoFallthrough(If.getThen());
+  bool ElseFallthrough = not isIndirectlyNoFallthrough(If.getElse());
+
+  // If both branches fall through, neither can be hoisted.
+  if (ThenFallthrough and ElseFallthrough)
+    return std::nullopt;
+
+  // If exactly one branch falls through, that one is hoisted.
+  if (ThenFallthrough != ElseFallthrough)
+    return static_cast<HoistingTarget>(ElseFallthrough);
+
+  bool ThenHasDirectNoFallthrough = hasSingleNoFallthrough(If.getThen());
+  bool ElseHasDirectNoFallthrough = hasSingleNoFallthrough(If.getElse());
+
+  // If exactly one branch contains a single directly non-fallthrough operation,
+  // that one is hoisted.
+  if (ThenHasDirectNoFallthrough != ElseHasDirectNoFallthrough)
+    return static_cast<HoistingTarget>(ElseHasDirectNoFallthrough);
+
+  // Otherwise, the sizes of both branches are approximated, and a decision is
+  // made by comparing those approximations.
+  double ThenWeight = approximateRegionWeight(If.getThen());
+  double ElseWeight = approximateRegionWeight(If.getElse());
+
+  if (ThenWeight == ElseWeight)
+    return std::nullopt;
```

Comment:
Hoist the `Else` if their score is equal.

This will likely require a few tests to be changed.

Reply:
With such a simple example it doesn't seem to make much of a difference.

But if the then and the else are big, and contain nested code, hoisting is always preferable.

Please hoist the else in case of score parity.